rs/execution_environment/src/execution_environment.rs

```diff
@@ -3150,6 +3152,48 @@ impl ExecutionEnvironment {
         guard.paused_install_code.remove(&id)
     }
 
+    fn abort_paused_executions_and_return_tasks(
+        &self,
+        paused_tasks: Vec<&ExecutionTask>,
+        log: &ReplicaLogger,
+    ) -> HashMap<PausedExecutionId, ExecutionTask> {
+        paused_tasks
+            .into_iter()
+            .map(|paused_task| match *paused_task {
+                ExecutionTask::PausedExecution { id, .. } => {
+                    let paused = self.take_paused_execution(id).unwrap();
+                    let (input, prepaid_execution_cycles) = paused.abort(log);
+                    (
+                        id,
+                        ExecutionTask::AbortedExecution {
+                            input,
+                            prepaid_execution_cycles,
+                        },
+                    )
+                }
+                ExecutionTask::PausedInstallCode(id) => {
+                    let paused = self.take_paused_install_code(id).unwrap();
+                    let (message, call_id, prepaid_execution_cycles) = paused.abort(log);
+                    (
+                        id,
+                        ExecutionTask::AbortedInstallCode {
+                            message,
+                            call_id,
+                            prepaid_execution_cycles,
+                        },
+                    )
+                }
+                ExecutionTask::AbortedExecution { .. }
+                | ExecutionTask::AbortedInstallCode { .. }
+                | ExecutionTask::Heartbeat
+                | ExecutionTask::GlobalTimer
+                | ExecutionTask::OnLowWasmMemory(..) => {
+                    unreachable!("Here must be a paused task.")
```

Comment:
Let's leave a comment on the safety of this. It's only called in one place, right after taking all `paused` tasks from the task queue, so it cannot happen unless there's a bug.

Additionally, we should make sure there's some test that covers the case where someone would accidentally break this.

Reply:
I have updated the comment. Regarding the test, is it okay if I add it in the follow-up https://github.com/dfinity/ic/pull/1836 since this function will be refactored almost completely?

Reply:
By refactoring you mean that the `unreachable` might be removed? I want to ensure that we cover potentially a case where the replica can panic with a test to avoid surprises. If this case can be triggered after we merge and until you refactor, there is a risk. How hard is it to have a test to be on the safe side?

Reply:
Another way to look at it: the assumption here is that `TaskQueue::get_all_paused_tasks()` returns only paused tasks and then this can never happen. If there was a bug in `get_all_paused_tasks()`, would any of our existing tests fail? i.e. would we catch if someone unintentionally changes the assumption? If yes, then that's also enough.